File: auto_download.js
const fetch = require('node-fetch');
const fs = require('fs');
const path = require('path');

// Function to download a single file
async function downloadFile(fileId, pattern, extension) {
    const url = `https://toto.dola/api/storage/substorage/${fileId}/download`;
    
    // Construct the filename using the provided pattern and extension
    const fileName = `${pattern}_${fileId}.${extension}`;
    const filePath = path.join(__dirname, fileName);  // Save in current directory
    
    try {
        // Fetch the file from the URL
        const response = await fetch(url);

        if (response.ok) {
            // Convert the response into a buffer (binary data)
            const buffer = await response.buffer();
            
            // Write the file to disk
            fs.writeFileSync(filePath, buffer);
            console.log(`Downloaded ${fileName} to ${filePath}`);
        } else {
            console.log(`Failed to download ${fileId}: ${response.status}`);
        }
    } catch (error) {
        console.log(`Error downloading ${fileId}: ${error.message}`);
    }
}

// Function to download multiple files
async function downloadFiles(fileIds, pattern, extension) {
    for (const fileId of fileIds) {
        await downloadFile(fileId, pattern, extension);
    }
}

// Example usage: Provide the list of file IDs, desired pattern, and extension
const fileIds = ["12345", "67890", "abcde"];  // Replace with your actual file IDs
const pattern = "customPattern";  // Input the desired pattern
const extension = "csv";  // Input the desired extension (e.g., csv, txt, pdf, etc.)

// Start downloading the files
downloadFiles(fileIds, pattern, extension);

File: auto_download.py
import aiohttp
import asyncio
import os

# Define the base URL and file IDs
BASE_URL = "https://toto.dola/api/storage/substorage/{file_id}/download"
OUTPUT_FOLDER = "./downloads"  # Change this to your preferred folder
PATTERN = "pattern_{file_id}.csv"

# Ensure the output directory exists
os.makedirs(OUTPUT_FOLDER, exist_ok=True)

async def download_file(session, file_id):
    """Download a file asynchronously given a file_id."""
    url = BASE_URL.format(file_id=file_id)
    filename = PATTERN.format(file_id=file_id)
    filepath = os.path.join(OUTPUT_FOLDER, filename)

    try:
        async with session.get(url) as response:
            if response.status == 200:
                with open(filepath, "wb") as f:
                    f.write(await response.read())  # Save the file
                print(f"✅ Downloaded: {filename}")
            else:
                print(f"❌ Failed: {filename} (HTTP {response.status})")
    except Exception as e:
        print(f"⚠️ Error downloading {filename}: {e}")

async def main(file_ids):
    """Download all files concurrently."""
    async with aiohttp.ClientSession() as session:
        tasks = [download_file(session, file_id) for file_id in file_ids]
        await asyncio.gather(*tasks)  # Run downloads in parallel

# List of file IDs to download
file_ids = ["12345", "67890", "abcde"]  # Replace with your actual IDs

# Run the async download script
asyncio.run(main(file_ids))
